Allow users without admin privileges open storage device on macOS

IgorAnohin opened the pull request:
**GoodDay:** [Flasher for Mac Can't Connect to M2](https://www.goodday.work/t/ehulLE)

## Motivation
Non-admin users are unable to reflash devices, because `sudo --askpass` doesn't work and they can't open storage devices for writing.
In general, there are 2 problems. Users without admin privileges:
1) Unable to unmount disk mount points.
2) Unable to open storage devices

## Solution
For the "1)" problem solution is [DADiskUnmount](https://developer.apple.com/documentation/diskarbitration/1492758-dadiskunmount?language=objc) function, which allows to unmout the whole disk. It's enough to be able to write.
An example was picked from https://github.com/raspberrypi/rpi-imager/blob/8a44c6b3ca99b87896451d7fe48342955675bad4/dependencies/mountutils/src/darwin/functions.cpp#L171


For the "2)" problem solution is `/usr/libexec/authopen` util, which allows entering admin password and get a special fd for storage file.
An example was picked from https://github.com/raspberrypi/rpi-imager/blob/8a44c6b3ca99b87896451d7fe48342955675bad4/mac/macfile.cpp#L25

## How to test
1) Get a storage device, that has mount points. For example, I picked ReachRS2.
2) Put the device into storage mode
3) Patch `devlib_cli` example. Here is the patch: 
[changed_devlib_cli.txt](https://github.com/emlid/devlib/files/5844038/changed_devlib_cli.txt)

> To apply the patch use the following command: `$ git apply changed_devlib_cli.txt`
4) Run `devlib_cli` and check the result. You should see something like:
```shell
$  ./devlib_cli
...output...
-> Use device on "20-3.4" port
...output...
-> macx_native: Unmounting
-> macx_native: Schedule session on run loop
-> macx_native: Waiting run loop
-> macx_native: Unmount callback
-> macx_native: Unmount success
...output...
-> Opening: false
...output...
-> Auth Opening: true
``` 

## Changes
- `devlib/native`:
    - `umount` -> `umountParition`
    - `umountDisk` function has been added and implemented on macOS
    - `authOpen` function has been added and implemented on macOS
- `StorageDeviceFile`:
    - `authOpen` function has been added
    - `StorageDeviceFileImpl` tries unmount the whole disk instead of mountpoints
    - unused 'mntptLock' object member has been removed from `StorageDeviceFileImpl`
- `devlib_cli`:
    - `QCoreApplication` events loop has been removed

IgorAnohin committed:
examples: devlivb_cli: remove QCoreApplication so the app doesn't hang

QCoreApplication.exec() function starts the main thread's event loop. Unfortunately, we don't stop the event loop, so the application will run forever.

IgorAnohin committed:
devlib: StorageDeviceFileImpl: remove unused 'mntptLock' object member

mntptLock is a private member that is never read.

IgorAnohin committed:
devlib: native: rename umount -> umountParition

In fact, the "umount" function will unmount only one partition. It would be nice to emphasize this in the name of the function.
Moreover, a new function will be created to unmount the entire disk.

IgorAnohin committed:
devlib: StorageDeviceFileImpl: try unmount disk instead of mountpoints

macOS platform does not allow users without admin privileges to unmount partitions separately. So, we should try to unmount the entire disk before the partitions.

IgorAnohin committed:
StorageDeviceFile: add function to open file with admin authorization

On macOS, even non-admin user can try to open protected file with administrator rights if it knows the password. It might be the only way to non-admin users open storage devices.
So, "authOpen" function has been added to StorageDeviceFile API.

IgorAnohin committed:
devlib: StorageDeviceFileImpl: remove redundant virtual keyword in funcs

"virtual" is redundant since functions are already declared as "override" .

smirnov1993 commented on devlib/impl/StorageDeviceFileImpl.cpp:
```diff
@@ -1,6 +1,28 @@
+
+    bool umountDevice(const std::shared_ptr<devlib::IStorageDeviceInfo> deviceInfo)
+    {
+        if (devlib::native::umountDisk(deviceInfo->filePath())) {
+            return true;
+        }
+
+        auto mntpts = deviceInfo->mountpoints();
+        for (auto const & mntpt : mntpts) {
+            auto mntptLock = mntpt->umount();
+            if (!mntptLock->locked()) {
+                return false;
+            }
```
> macOS platform does not allow users without admin privileges to unmount partitions separately. So, we should try to unmount the entire disk before the partitions.

Does it mean that all users are able to unmount the entire disk and no one will ever reach `mntpt->umount()` command?

IgorAnohin replied:
> Does it mean that all users are able to unmount the entire disk?

In general, yes

> no one will ever reach mntpt->umount() command

I think it's possible if `unmountDisk` will fail for some reasons. But during testing, I have not seen such situations.
This is why I left log messages. This will help us in debugging the code if users encounter something unusual.

ivnsmirnov reviewed (changes requested):
1) 1958a40944f2f171a36f71a47771a6b329f9c5c3 `devlib: native: add function to unmount the entire disk`
8ef853f84d839a955a0bba52f0f706362c34c94e `devlib: native: add function to open file with authentication`
You say that you add a function and even describe the idea, but nothing happens in the commit (just API appears). We can squash commits with the addition of API and implementation. 


2) 023266fc28de1d1cd2e8d99bea551329a90cf657 `devlib: native: macosx: implement umountDisk function`
a0179d7332e1e55af417b6109742ca3543a83e1f `devlib: native: macosx: implement authOpen function` 
The code from rpi must be provided with copyright notice and the list of changes made to the  original code. Move the rpi code to the separate file and expand the copyright (check this [example](https://github.com/emlid/reach-ublox-daemon/blob/76863ba1a8d086385b5f8a07c4a15408a3ed379f/include/drivers/ublox.h#L1-L38) . The part with changes might be described with more details). 
These files better be added in separate commit.

GlebIgnatevEmlid commented on devlib/native/macosx_native.cpp:
```diff
@@ -163,6 +166,258 @@ namespace macxutil {
+        AuthorizationRights rights = {1, &item};
+        AuthorizationFlags flags = kAuthorizationFlagInteractionAllowed | kAuthorizationFlagExtendRights
+                                   | kAuthorizationFlagPreAuthorize;
+        AuthorizationRef authRef;
```
Could be wrapped as well calling `AuthorizationFree` in the deleter.

IgorAnohin replied:
Applied

GlebIgnatevEmlid commented on devlib/native/macosx_native.cpp:
```diff
@@ -163,6 +166,258 @@ namespace macxutil {
+        qCDebug(macxutil::macxlog()) << "Creating DA session";
+        DASessionRef session = DASessionCreate(kCFAllocatorDefault);
+
+        if (session == NULL) {
```
`nullptr`?

IgorAnohin replied:
Fixed

GlebIgnatevEmlid commented on devlib/native/macosx_native.cpp:
```diff
@@ -163,6 +166,258 @@ namespace macxutil {
+
+        const char * cmd = "/usr/libexec/authopen";
+        QByteArray mode = QByteArray::number(O_RDWR);
+        ::execl(cmd, cmd, "-stdoutpipe", "-extauth", "-o", mode.data(), filename, NULL);
```
`nullptr`?

IgorAnohin replied:
Applied

GlebIgnatevEmlid reviewed (changes requested):
Good job! Works as expected. I agree with @smirnov1993 regarding commits squashing/separation. I've left a few comments.

IgorAnohin commented on devlib/native/macosx_native.cpp:
```diff
@@ -163,6 +166,258 @@ namespace macxutil {
+
+        const char * cmd = "/usr/libexec/authopen";
+        QByteArray mode = QByteArray::number(O_RDWR);
+        ::execl(cmd, cmd, "-stdoutpipe", "-extauth", "-o", mode.data(), filename, NULL);
+
+        // Exit with error code -1 in case of execl error
+        ::exit(-1);
```
This code looks like an `else` situation is possible, doesn't it?

GlebIgnatevEmlid replied:
Otherwise, it looks like `exit(-1)` should be executed all the time, whereas in fact it's unreachable in case `execl` succeeds.

IgorAnohin committed:
native: macos: rename macxutil namespace to macx_utils

IgorAnohin committed:
native: macos: move utils to separate file

IgorAnohin commented on devlib/native/linux_native.cpp:
```diff
@@ -118,6 +118,13 @@ auto devlib::native::umountPartition(QString const& mntpt)
 }
 
 
+// Temporary unsupported
```
Fixed

IgorAnohin commented on devlib/native/macosx_native.cpp:
```diff
@@ -163,6 +164,129 @@ namespace macxutil {
+
+        // Get a disk object from the disk path
+        qCDebug(macxutil::macxlog()) << "Getting disk object";
+        DADiskRef disk = DADiskCreateFromBSDName(kCFAllocatorDefault, session, device);
```
Thanks. Fixed

IgorAnohin commented on devlib/native/macosx_native.cpp:
```diff
@@ -163,6 +164,129 @@ namespace macxutil {
+            context.result = UnmountResult::RunloopStallError;
+        }
+
+        // Clean up the session
+        qCDebug(macxutil::macxlog()) << "Releasing session & disk object";
+        DASessionUnscheduleFromRunLoop(session, CFRunLoopGetCurrent(), kCFRunLoopDefaultMode);
+        CFRelease(session);
```
`session` and `disk` have been wrapped into `unique_ptr`

IgorAnohin committed:
native: macos: add unmountDiskWithRunLoop func to unmount entire disk

On macOS, users without administrator rights cannot unmount partitions. At the same time, the user can unmount the entire disk.
So, "unmountDiskWithRunLoop" function has been added.

The code is based on https://github.com/raspberrypi/rpi-imager/blob/qml/dependencies/mountutils/src/darwin/functions.cpp#L171
It uses the Apple DADiskUnmount function, which must be run inside a RunLoop because it's asynchronous.

IgorAnohin committed:
devlib: native: add umountDisk func to unmount the entire disk

On macOS, users without administrator rights cannot unmount partitions. At the same time, the user can unmount the entire disk.
So, "umountDisk" function has been added to the native module API.

IgorAnohin committed:
native: macos: add authOpenStorageDevice to open dev with authentication

On macOS, even non-admin user can try to open protected file with administrator rights if it knows the password. It might be the only way to non-admin users open storage devices.
So, "authOpenStorageDevice" function has been added.

The code is based on https://github.com/raspberrypi/rpi-imager/blob/qml/mac/macfile.cpp#L25
It uses built-in "/usr/libexec/authopen" command, which returns special fd for a file.

IgorAnohin committed:
devlib: native: add authOpen func to open file with authentication

On macOS, even non-admin user can try to open protected file with administrator rights if it knows the password. It might be the only way to non-admin users open storage devices.
So, "authOpen" function has been added to the native module API.

IgorAnohin commented:
> ... Move the rpi code to the separate file and expand the copyright ...

RPI-Imager code has been moved to separate files. Copyright has been expanded.

> The part with changes might be described with more details).

I didn't find how to explain "changes", because I only refactored code a little bit. Some methods have been renamed and rearranged. The same has been done for variables.
There were no functional changes. 

>     These files better be added in separate commit.
402fbf70caac9379c220a325fd1ccb2e568a631e and 6b4868c6284b7a0f6abe6d30e72c81ed04bb14f6

GlebIgnatevEmlid reviewed (approved):
Nice refactoring, looks much better without this 700+ LoC file.
